### PR TITLE
Custom progress bar

### PR DESCRIPTION
**Pull request by anisehsani**
Implements custom bar format change
Allows user to pass in a `dataloader_label_to_metrics` dictionary that dictates what metrics are returned for each progress bar.
### Example 1
```yaml
train_dataset:
  &train_dataset
  mnist:
    use_synthetic: true
evaluators:
  eval1:
    label: eval/batch
    eval_interval: 1ba
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
  eval2:
    label: eval/ep
    eval_interval: 1ep
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
model:
  mnist_classifier:
    num_classes: 10
max_duration: 6000sp
train_batch_size: 2000
eval_batch_size: 2000
dataloader:
  pin_memory: true
  timeout: 0
  prefetch_factor: 2
  persistent_workers: true
  num_workers: 1
train_subset_num_batches: 2

progress_bar: False
loggers:
  - progress_bar:
      dataloader_label_to_metrics: {'eval/batch': False, 'eval/ep': 'dataloader_label={state._dataloader_label}, accuracy={Accuracy}', 'train': 'custom_message, loss={state.loss}'}
      stream: stderr
```

```
eval/batch     Sample  2000:  100%|█████████████████████████| 2/2 [00:00<00:00, 36.79ba/s]                                                                                                                                          
eval/batch     Sample  4000:  100%|█████████████████████████| 2/2 [00:00<00:00, 45.81ba/s]                                                                                                                                          
eval/ep        Sample  4000:  100%|█████████████████████████| 2/2 [00:00<00:00, 40.94ba/s, dataloader_label="eval/ep", accuracy=0.1213]                                                                                             
eval/batch     Sample  6000:  100%|█████████████████████████| 2/2 [00:00<00:00, 43.85ba/s]                                                                                                                                          
eval/ep        Sample  6000:  100%|█████████████████████████| 2/2 [00:00<00:00, 44.50ba/s, dataloader_label="eval/ep", accuracy=0.1240]                                                                                             
train                         100%|█████████████████████████| 6000/6000 [00:01<00:00, 4419.43sp/s, custom_message, loss=149690.7500]    
```

### Example 2
```yaml
train_dataset:
  &train_dataset
  mnist:
    use_synthetic: true
evaluators:
  eval1:
    label: eval/batch
    eval_interval: 1ba
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
  eval2:
    label: eval/ep
    eval_interval: 1ep
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
model:
  mnist_classifier:
    num_classes: 10
max_duration: 2ep
train_batch_size: 2000
eval_batch_size: 2000
dataloader:
  pin_memory: true
  timeout: 0
  prefetch_factor: 2
  persistent_workers: true
  num_workers: 1
train_subset_num_batches: 2
progress_bar:True
```

```

eval/batch     Batch   1:  100%|█████████████████████████| 2/2 [00:00<00:00, 37.22ba/s, CrossEntropy=2.2993, Accuracy=0.1482]                                                                                                       
eval/batch     Batch   2:  100%|█████████████████████████| 2/2 [00:00<00:00, 45.15ba/s, CrossEntropy=2.2981, Accuracy=0.1325]                                                                                                       
train          Epoch   0:  100%|█████████████████████████| 2/2 [00:01<00:00,  1.38s/ba, loss/train=151222.2031]                                                                                                                     
eval/ep        Epoch   0:  100%|█████████████████████████| 2/2 [00:00<00:00, 46.41ba/s, CrossEntropy=2.3046, Accuracy=0.1028]                                                                                                       
eval/batch     Batch   3:  100%|█████████████████████████| 2/2 [00:00<00:00, 41.18ba/s, CrossEntropy=2.2968, Accuracy=0.1405]                                                                                                       
eval/batch     Batch   4:  100%|█████████████████████████| 2/2 [00:00<00:00, 13.85ba/s, CrossEntropy=2.2968, Accuracy=0.1308]                                                                                                       
train          Epoch   1:  100%|█████████████████████████| 2/2 [00:00<00:00, 19.56ba/s, loss/train=150556.0312]                                                                                                                     
eval/ep        Epoch   1:  100%|█████████████████████████| 2/2 [00:00<00:00, 45.09ba/s, CrossEntropy=2.3079, Accuracy=0.0970]     
```

Changing the progress bar section of the yaml to this gives the same results:
```yaml
  - progress_bar:
      dataloader_label_to_metrics: {}
      stream: stderr
```

### CODE REVIEW

File: composer/loggers/progress_bar_logger.py [hunk 1]
```diff
@@ -19,7 +19,14 @@
 
 __all__ = ['ProgressBarLogger']
 
-_IS_TRAIN_TO_KEYS_TO_LOG = {True: ['loss/train'], False: ['metrics/eval/Accuracy']}
+
+class FormatDict(dict):
```
Comment (ravi-mosaicml) [inline composer/loggers/progress_bar_logger.py:23]: Hmm, I would prefer not to subclass dict. Could we replace this with a function that takes a dict and returns a dict?

File: composer/loggers/progress_bar_logger.py [hunk 2]
```diff
@@ -56,10 +65,21 @@ def __init__(
...
+            curr_metrics = state.current_metrics.get(dataloader_label, {})
+            for current_metric, v in curr_metrics.items():
+                formatted_data[current_metric] = format_log_data_value(v)
+            if self.is_train and state.timestamp.batch_in_epoch != 0:
+                formatted_data['loss/train'] = format_log_data_value(state.loss)
+            self.pbar.set_postfix(formatted_data)
+        else:
+            current_metrics = state.current_metrics.get(dataloader_label, {})
+            metric_string = self.metric_string.format_map(FormatDict(current_metrics))
+            metric_string = metric_string.format_map(FormatDict(state.__dict__))
```
Comment (ravi-mosaicml) [inline composer/loggers/progress_bar_logger.py:81]: Does `state.__dict__` work with state elements defined as `@property`?

File: composer/loggers/progress_bar_logger.py [hunk 3]
```diff
@@ -56,10 +65,21 @@ def __init__(
...
 
-    def log_data(self, data: Dict[str, Any]):
-        formatted_data = {k: format_log_data_value(v) for (k, v) in data.items() if k in self.keys_to_log}
-        self.metrics.update(formatted_data)
-        self.pbar.set_postfix(self.metrics)
+    def log_data(self, state: State):
+        dataloader_label = state.dataloader_label
+        if self.metric_string is None:
+            formatted_data = {}
+            curr_metrics = state.current_metrics.get(dataloader_label, {})
```
Comment (ravi-mosaicml) [inline composer/loggers/progress_bar_logger.py:72]: Could we unify the logic between the if and the else -- e.g. `if metric_string is None: metric_string = ', '.join('k={k}' for k in curr_metrics)`, and then use the same logic for both cases?

**Review by ravi-mosaicml — COMMENTED** [review]
See comments -- will give a closer review once pr1 is refactored / merged

File: composer/loggers/progress_bar_logger.py (hunk 2 above)
Comment (anisehsani) [reply composer/loggers/progress_bar_logger.py]: Yes and no. For example it won't work for `dataloader_label`, but it will work for `_dataloader_label`. I can mention this in the doc strings.

File: composer/loggers/progress_bar_logger.py (hunk 3 above)
Comment (anisehsani) [reply composer/loggers/progress_bar_logger.py]: I don't think that would reduce the complexity much because of the default keys I need to add, like `'loss_train`' or the default accuracy

File: composer/trainer/trainer.py [hunk 4]
```diff
@@ -838,12 +838,18 @@ def __init__(
...
         loggers = list(ensure_tuple(loggers))
         if any(isinstance(x, ProgressBarLogger) for x in loggers):
-            warnings.warn(
-                DeprecationWarning(
-                    (f'Specifying the {ProgressBarLogger.__name__} via `loggers` is deprecated. Instead, '
-                     'please specify `progress_bar`, `log_to_console`, `log_level`, and `stream` arguments when '
-                     'constructing the trainer. If specified, these arguments will be ignored, as the '
-                     f'{ProgressBarLogger.__name__} was already created.')))
+            if log_to_console is not None:
+                raise ValueError(
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:842]: Since we only have one `ProgressBarLogger` instance now, I don't think we need these changes anymore...

**Review by ravi-mosaicml — COMMENTED** [review]
Sorry for the delay on the review. I think the main theme of comments is simplifying the logic in the `_ProgressBar` class. I didn't give the test cases or docstrings a close review, but will do that next time.

File: composer/trainer/trainer.py (hunk 4 above)
Comment (anisehsani) [reply composer/trainer/trainer.py:842]: I think we still do right? Aren't these for the default progress bar? Won't the user might want to construct the `dataloader_label_to_metrics` dict, which gets passed into the logger object. Should I make the `dataloader_label_to_metrics` a trainer parameter instead?

File: composer/loggers/progress_bar_logger.py [hunk 5]
```diff
@@ -19,7 +20,18 @@
...
 
-_IS_TRAIN_TO_KEYS_TO_LOG = {True: ['loss/train'], False: ['metrics/eval/Accuracy']}
+
+class MetricFormatter(string.Formatter):
+
+    def get_field(self, field_name: str, args: Sequence[Any], kwargs: Mapping[str, Any]) -> Any:
+        try:
+            obj, arg_used = super().get_field(field_name, args, kwargs)
+            return format_log_data_value(obj), arg_used
+        except (AttributeError, KeyError):
```
[the remark this answers is not in the document]
Comment (anisehsani) [reply composer/loggers/progress_bar_logger.py:30]: We don't need to - can change

File: composer/loggers/progress_bar_logger.py [hunk 6]
```diff
@@ -30,17 +42,17 @@ def __init__(
         position: int,
         bar_format: str,
         file: TextIO,
-        metrics: Dict[str, Any],
-        keys_to_log: List[str],
         timestamp_key: str,
+        dataloader_label_to_metrics: Optional[Dict[str, Union[str, bool]]] = None,
+        is_train: bool = False,
```
[the remark this answers is not in the document]
Comment (anisehsani) [reply composer/loggers/progress_bar_logger.py:47]: I can change the metrics part. I would like to keep the `is_train` variable since that indicates to log the default `'loss/train' metric. In the previous implementation, there was also a similar variable, it was just a part of the `keys_to_log` field.

File: composer/loggers/progress_bar_logger.py [hunk 7]
```diff
@@ -52,14 +64,30 @@ def __init__(
...
+        dataloader_label = state.dataloader_label
+        assert dataloader_label is not None
+        if self.dataloader_label_to_metrics is not None and dataloader_label in self.dataloader_label_to_metrics:
+            entry = self.dataloader_label_to_metrics[dataloader_label]
+            if isinstance(entry, str):
+                metric_string = entry
+                metrics = state.current_metrics.get(dataloader_label, {})
+                ft = MetricFormatter()
+                metric_string = ft.format(metric_string, state=state)
+                metric_string = ft.format(metric_string, **metrics)
```
[the remark this answers is not in the document]
Comment (anisehsani) [reply composer/loggers/progress_bar_logger.py:81]: No, we don't, it was just clearer for me to understand what was happening and for debugging

File: composer/trainer/trainer.py (hunk 4 above)
Comment (ravi-mosaicml) [reply composer/trainer/trainer.py:842]: Yes, let's also propagate this parameter through the trainer and trainer hparams.
